### PR TITLE
Enable LPSPI RTIO Support

### PR DESCRIPTION
**Pull request by EmilioCBen**
Tested with the mimxrt1170_evk_cm7 using the tests/spi/spi_loopback test.

### CODE REVIEW

**Comment by bperseghetti** [comment]
Probably should also add the VMU_RT1170 to the test.

File: drivers/spi/spi_mcux_lpspi.c
Comment (teburd) [inline drivers/spi/spi_mcux_lpspi.c:684]: when would this not be 0? It seems like this implies a blocking flow at all times.

What should happen is the transceive above should be asynchronously done and upon completion the next transceive started.

Or maybe better is the entire transaction is loaded into a hardware queue and started asynchronously, then upon completion things are marked as such.

**Review by teburd — COMMENTED** [review]
Unclear if this works in an asynchronous (event driven) way as would be expected, some clarification is needed

File: drivers/spi/spi_mcux_lpspi.c
Comment (decsny) [reply drivers/spi/spi_mcux_lpspi.c:684]: yes, driver should enable dma if using rtio and call transceive_dma here instead of regular blocking transceive

File: drivers/spi/Kconfig.mcux_lpspi
Comment (decsny) [inline drivers/spi/Kconfig.mcux_lpspi:19]: add also `default y if SPI_RTIO`

Comment (decsny) [reply drivers/spi/Kconfig.mcux_lpspi:19]: and maybe `select SPI_ASYNC if SPI_RTIO`

File: drivers/spi/spi_mcux_lpspi.c
Comment (tbursztyka) [inline drivers/spi/spi_mcux_lpspi.c:517]: Why these switch on/off of the CS line? 

and why the spi_mcux_configure() call, since it's already called below at line 520 ?

Fidally, how these changes fit with RTIO addition in this driver?

**Review by tbursztyka — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_mcux_lpspi.c
Comment (teburd) [reply drivers/spi/spi_mcux_lpspi.c:684]: I’m working on adding polling support as well, so if for small transfers it makes more sense to have a tight poll loop, rtio will be able to support that. It should fix the Sam spi issue as well.

**Comment by EmilioCBen** [comment]
> Probably should also add the VMU_RT1170 to the test.

I have now added this to the testcase.yaml

File: drivers/spi/spi_mcux_lpspi.c
Comment (EmilioCBen) [reply drivers/spi/spi_mcux_lpspi.c:506]: Thank you, I don't know where this came from, I have removed this.

File: drivers/spi/spi_mcux_lpspi.c
Comment (teburd) [inline drivers/spi/spi_mcux_lpspi.c:677]: next needs to be serialized in a spin lock as it is in the sam driver

It also needs to verify there's no ongoing current work before popping and starting the next thing on the queue.

**Review by teburd — CHANGES_REQUESTED** [review]
There's a race in the submit call that needs correcting

Please carefully look over the way submit is written for the sam spi driver and replicate the structure there for now.

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/spi/spi_sam.c#L685

Note that a spin lock is taken (serialized single consumer of the request queue), checks if there's ongoing work. If no work is ongoing then pop and start, otherwise do nothing.

File: drivers/spi/spi_mcux_lpspi.c
Comment (EmilioCBen) [reply drivers/spi/spi_mcux_lpspi.c:677]: I wondered about that, but when the LPSPI driver reaches the transceive function, would this not stop the race condition?
https://github.com/zephyrproject-rtos/zephyr/blob/9686f6e000c746538ec7ab78f6d8abf299d24e95/drivers/spi/spi_mcux_lpspi.c#L488C1-L488C1

File: drivers/spi/spi_mcux_lpspi.c
Comment (decsny) [reply drivers/spi/spi_mcux_lpspi.c:506]: FYI you just pushed and it's not removed dont know if it's intentional

File: drivers/spi/spi_mcux_lpspi.c
Comment (decsny) [reply drivers/spi/spi_mcux_lpspi.c:677]: > I wondered about that, but when the LPSPI driver reaches the transceive function, would this not stop the race condition?

looks like it would not, data is already changed before getting to this

File: drivers/spi/spi_mcux_lpspi.c
Comment (teburd) [inline drivers/spi/spi_mcux_lpspi.c:627]: This seems to be mixing transmit and receive in the same spi_buf setup which looks a bit wrong to me

File: drivers/spi/spi_mcux_lpspi.c
Comment (EmilioCBen) [reply drivers/spi/spi_mcux_lpspi.c]: Thank you again, it is now removed.

File: drivers/spi/spi_mcux_lpspi.c
Comment (EmilioCBen) [reply drivers/spi/spi_mcux_lpspi.c:708]: Thank you both, I have fixed this and there should no longer be a race condition.

**Comment by EmilioCBen** [comment]
@teburd @tbursztyka 
Would you two mind giving this PR another look please.

File: tests/drivers/spi/spi_loopback/boards/mimxrt1170_evk_cm7.conf
Comment (EmilioCBen) [reply tests/drivers/spi/spi_loopback/boards/mimxrt1170_evk_cm7.conf]: Fixed

File: drivers/spi/spi_mcux_lpspi.c
Comment (EmilioCBen) [reply drivers/spi/spi_mcux_lpspi.c]: Fixed

File: drivers/spi/spi_mcux_lpspi.c
Comment (EmilioCBen) [reply drivers/spi/spi_mcux_lpspi.c]: This should no longer be the case.

**Review by danieldegrasse — DISMISSED** [review]

**Review by teburd — DISMISSED** [review]
Much nicer, thank you!

I'm looking forward to seeing what can be done with this on NXP peripherals as I believe there's better support for command queuing in both the IP block and eDMA.

**Review by bperseghetti — DISMISSED** [review]
Thanks Emilio, looks good.

File: drivers/spi/spi_mcux_lpspi.c
Comment (teburd) [inline drivers/spi/spi_mcux_lpspi.c:691]: rtio_iodev_sqe_err(txn_head, -EIO); or something likely needed here

**Review by MaureenHelm — DISMISSED** [review]
Very cool!

File: drivers/spi/spi_mcux_lpspi.c
Comment (EmilioCBen) [reply drivers/spi/spi_mcux_lpspi.c:691]: understood, added this call.

File: drivers/spi/Kconfig.mcux_lpspi
Comment (tbursztyka) [inline drivers/spi/Kconfig.mcux_lpspi:23]: you could have a "depends on SPI_RTIO instead of such if/endif.

Comment (decsny) [reply drivers/spi/Kconfig.mcux_lpspi:23]: the if/endif is more scalable syntax

Comment (EmilioCBen) [reply drivers/spi/Kconfig.mcux_lpspi:23]: Can you help me understand the advantage of "depends on" over "if/endif"?
Perhaps I missed something here.

**Comment by teburd** [comment]
@tbursztyka any remaining issues still? it would be nice to have at least two parts supporting this API to further validate it, and I'd love to use something other than a sam e70 :-) Thanks!

**Comment by dleach02** [comment]
@tbursztyka, can you return to this PR to see if @EmilioCBen has addressed your change request? We would like to get this merged.

**Review by bperseghetti — APPROVED** [review]

**Review by teburd — APPROVED** [review]
Needs maintainer approval but refreshing my +1

**Comment by dleach02** [comment]
@tbursztyka please review the changes to determine if your requests have been dealt with

**Review by decsny — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]

**Review by tbursztyka — APPROVED** [review]

**Pull request merged**
